### PR TITLE
[IMP] sale_coupon_multi_use: m2m relation

### PR DESCRIPTION
**Pull request by oerp-odoo**
`coupon_multi_use_ids` on sale.order is now used to track multi-use
coupons on sale orders. `applied_coupon_ids` field did not suit multi-use
coupons case, because it is o2m field where if same coupon is applied
on next sale order, its relation with old sale order is removed.

This relation follows similar logis as is for `applied_coupon_ids` field
relation logic.

### CODE REVIEW

File: sale_coupon_multi_use/i18n/fr.po
Comment (jcoux) [inline sale_coupon_multi_use/i18n/fr.po]:
```suggestion
msgstr "Appliqué sur les bons de commandes"
```

File: sale_coupon_multi_use/i18n/fr.po
Comment (jcoux) [inline sale_coupon_multi_use/i18n/fr.po]:
```suggestion
msgstr "Le coupon a déjà été appliqué sur la même récompense"
```

File: sale_coupon_multi_use/i18n/fr.po
Comment (jcoux) [inline sale_coupon_multi_use/i18n/fr.po]:
```suggestion
msgstr "Code du bon de réduction"
```

File: sale_coupon_multi_use/models/sale_coupon.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_coupon.py]: You don't need to load `coupon_multi_use_ids` to use the mapped function.
Let Odoo manage the mapped entirely.
```suggestion
        if self.program_id in order.mapped("coupon_multi_use_ids.program_id"):
```

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]:
```suggestion
        multi_use_coupons = (
            self.coupon_multi_use_ids - self.applied_coupon_ids
        )
```

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: Remove this useless function used only once time and with a parameter never used.

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: Remove useless `filter_func` parameter.

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: Use a function on sale coupon to delegate the state management, by example `use_coupon()`

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: Use a function on sale coupon to delegate the state management, by example `reset_coupon_state()`

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: Use a function on sale coupon to delegate the state management, by example `use_coupon()`

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: No need to test `coupons_multi_use`.
Let Odoo manage it in `write` function.
```suggestion
            order.with_context(**{ORDER_CTX_KEY: order}).coupon_multi_use_ids.write(
                {"state": "used"}
            )
```

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: A comment on this function is welcome.

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: Remove this filter function and copy/paste the code in following filtered.
It's not easy to read like this, it will better to read/compare with your following code:
```
related_program = self.env["sale.coupon.program"].search(
    [("discount_line_product_id", "=", line.product_id.id)]
)
```
and
```
                related_program_lines |= (
                    order.order_line.filtered(
                        lambda l: l.product_id
                        == related_program.discount_line_product_id
                    )
                    - line
```

**Review by jcoux — CHANGES_REQUESTED** [review]

File: sale_coupon_multi_use/models/sale_coupon.py
Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_coupon.py]: Is there a difference?

File: sale_coupon_multi_use/models/sale_order.py
Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: It will call write with empty recordset, which is ugly.

File: sale_coupon_multi_use/models/sale_coupon.py
Comment (jcoux) [reply sale_coupon_multi_use/models/sale_coupon.py]: Yes, you let Odoo manage data loading.
In your case, you load all `coupon_multi_use_ids` to access to programs.
In the full mapped, you delegate this to Odoo, and maybe it's better optimized, maybe not, but it's the responsibility of Odoo, and if in a future patch they optimize or changed something in the mapped implementation, you will benefit of it.

Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_coupon.py]: Hm, but this was copy/pasted from this `self.program_id in order.applied_coupon_ids.mapped('program_id'):` (original check) Just changed field used, but case is basically the same.

File: sale_coupon_multi_use/models/sale_order.py
Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: Yes, that param is leftover from previous iteration. It can be removed. Though function usefulness should not be defined how many times its called. Its other purpose is SoC. Though I guess in this case, we can remove it.

File: sale_coupon_multi_use/models/sale_order.py
Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: Leftover

File: sale_coupon_multi_use/models/sale_coupon.py
Comment (jcoux) [reply sale_coupon_multi_use/models/sale_coupon.py]: The Odoo code is not always a good reference.

File: sale_coupon_multi_use/models/sale_order.py
Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: Don't get it what you intend to do here. Your showed code does different thing than mine. We need to filter coupons, not related program lines.

Comment (jcoux) [reply sale_coupon_multi_use/models/sale_order.py]: * Remove completely the function `_get_order_line_coupons_filter` (lines 6-7).
* Rewrite line 108:
`coupons_to_detach = order.coupon_multi_use_ids.filtered(lambda r: r.program_id.discount_line_product_id == line.product_id)`
* Rewrite lines 112-114:
`coupons_to_reactivate = (coupons_to_detach - order.applied_coupon_ids.filtered(lambda r: r.program_id.discount_line_product_id == line.product_id))`

Yes, you copy/paste two times the lambda filter, but the code will be more easy to read.

File: sale_coupon_multi_use/models/sale_order.py
Comment (jcoux) [inline sale_coupon_multi_use/models/sale_order.py]: For me all these 3 functions:
- _get_coupons_from_2many_commands
- _get_multi_use_coupons_from_2many_commands
- write

are too complex compared to the needs.

I think we can do more easy:
```
    def write(self, vals):
        """Extend to add multi-use coupons."""
        result = super().write(vals)
        if vals.get("applied_coupon_ids"):
            for order in self:
                order.coupon_multi_use_ids = order.applied_coupon_ids.filtered("multi_use")
        return result
```

**Review by jcoux — CHANGES_REQUESTED** [review]
Thanks for changes.
Last comment for me.

File: sale_coupon_multi_use/models/sale_order.py
Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: This calls write twice. Which is bad.

Comment (jcoux) [reply sale_coupon_multi_use/models/sale_order.py]: It's not bad if the final result is very more easy, then more easy to read/maintainable.

Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: Hm, I checked your suggestion again and actually it would work incorrectly. It would cause same problem this PR was created to solve in a first place.

Thing is you need to check what is actually done with `applied_coupon_ids`. You can't assume correct command was used.

Now because `applied_coupon_ids` is o2m field, imagine this case:

SO1: we apply multi-use coupon on applied_coupon_ids (4 or 6 cmd used). It gets applied on coupon_multi_use_ids. So far so good.
Then we apply same multi-coupon on SO2 via applied_coupon_ids. Because that field its o2m field, relation will be lost on SO1.
Now let say we also apply single-use coupon on SO1. It gets applied on applied_coupon_ids and triggers part of code you suggested. Now it finds no multi-use coupons on applied_coupon_ids, thus overwrites all multi-use coupons and causes same problem we have without this PR.

Well I guess we could then only add new ones on coupon_multi_use_ids, but I wonder if then its any easier to read (though I dont really find my part hard to read, but thats my opinion:)).

Comment (jcoux) [reply sale_coupon_multi_use/models/sale_order.py]: Ok I understand, then 2 things:

- Add all this explanation into the code to help people to correctly understand what you do and why you do it like this.
- `applied_coupon_ids` must always be empty when we have multi use module installed. Because the value of `applied_coupon_ids` is wrong, we don't want to keep it.

Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: applied_coupon_ids can't be kept empty, because it is also used for single use coupon (native functionality).
Though I guess what could be done, is to not apply multi-coupon IDs on `applied_coupon_ids` and instead of copying those ids, move them. So then it would simplify relation management (we would only check applied_coupon_ids once and on other parts deal with coupon_multi_use_ids)

Comment (jcoux) [reply sale_coupon_multi_use/models/sale_order.py]: :+1: with your proposition! :)

Comment (oerp-odoo) [reply sale_coupon_multi_use/models/sale_order.py]: @jcoux updated. Moving 2many commands between keys got too complicated (before write), so refactored with multiple write. Though still needed to use commands there to make sure both fields are properly updated (because we need to be aware of specific 2many commands either way).

**Review by jcoux — APPROVED** [review]
Ok for me, thanks for changes!

Please don't merge without approval of another people to have a double validation.

**Review by i-vyshnevska — APPROVED** [review]

**Pull request merged**
